python/shotgun_desktop/startup.py

```diff
@@ -698,6 +699,45 @@ def get_bundle_cache_location(self):
             return os.environ.get("SHOTGUN_DESKTOP_BUNDLE_CACHE_LOCATION") or None
 
 
+def automatic_claims_renewal(user, preemtive_renewal_threshold=0.9):
+    """
+    Handles automatic renewal of the SAML2 claims for the user.
+
+    :params user: an already logged in user.
+    :params preemtive_renewal_threshold: How far into the claims duration we will attempt renewal.
+                                         Defaults to 90%, usually 3 minutes 45 seconds.
+
+    This will only be used in a context where SSO is enabled on the
+    Shotgun site.
+    """
+    from tank.authentication import interactive_authentication
+    from sgtk.authentication import AuthenticationCancelled
+
+    logger.debug("Automatic claims renewal")
+    try:
+        previous_expiration = user._impl.get_saml_expiration()
```

Comment:
I feel this automatic claims renewal is just an implementation detail of SSO. Why can't we have this be done automatically by Toolkit in the background? Is it because there's no way to stop it? Could we bind the lifescope of this timer to the life of an engine, like we do for the Totanto thread @manneohrstrom ?